WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,107 @@
-		private object TriggerLock { get; }
 		public TimeSpan Period { get; }
-		public Exception LastException { get; set; }
-		public long LastExceptionCount { get; set; }
-		public DateTimeOffset LastExceptionFirstAppeared { get; set; }
 
-		private void ResetLastException()
-		{
-			LastException = null;
-			LastExceptionCount = 0;
-			LastExceptionFirstAppeared = DateTimeOffset.MinValue;
-		}
+		private volatile TaskCompletionSource<bool>? _tcs;
+
+		private LastExceptionTracker ExceptionTracker { get; }
 
+		/// <summary>
+		/// Normally, <see cref="ActionAsync(CancellationToken)"/> user-action is called every time that <see cref="Period"/> elapses.
+		/// This method allows to expedite the process by interrupting the waiting process.
+		/// </summary>
+		/// <remarks>
+		/// If <see cref="ExecuteAsync(CancellationToken)"/> is not actually in waiting phase, this method call makes
+		/// sure that next waiting process will be omitted altogether.
+		/// </remarks>
 		public void TriggerRound()
 		{
-			lock (TriggerLock)
-			{
-				TriggeringCts?.Cancel();
-			}
+			// Note: All members of TaskCompletionSource<TResult> are thread-safe and may be used from multiple threads concurrently.
+			_tcs?.TrySetResult(true);
 		}
 
+		/// <summary>
+		/// Abstract method that is called every <see cref="Period"/> or sooner when <see cref="TriggerRound"/> is called.
+		/// </summary>
 		protected abstract Task ActionAsync(CancellationToken cancel);
 
 		/// <inheritdoc />
 		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
 			while (!stoppingToken.IsCancellationRequested)
 			{
+				_tcs = new TaskCompletionSource<bool>();
+
 				try
 				{
+					// Do user action.
 					await ActionAsync(stoppingToken).ConfigureAwait(false);
-					LogAndResetLastExceptionIfNotNull();
+
+					// Log previous exception if any.
+					LogLastException(ExceptionTracker.LastException);
 				}
 				catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
 				{
 					Logger.LogTrace(ex);
 				}
 				catch (Exception ex)
 				{
-					// Only log one type of exception once.
-					if (LastException is null // If the exception never came.
-						|| ex.GetType() != LastException.GetType() // Or the exception have different type from previous exception.
-						|| ex.Message != LastException.Message) // Or the exception have different message from previous exception.
+					// Exception encountered, process it.
+					LogLastException(ExceptionTracker.Process(ex));
+
+					Logger.LogError(ExceptionTracker.LastException!.Exception);
+				}
+
+				// Wait for the next round.
+				try
+				{
+					using var cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+					var linkedTcs = _tcs; // Copy reference so it cannot change.
+
+					if (linkedTcs.Task == await Task.WhenAny(linkedTcs.Task, Task.Delay(Period, cts.Token)).ConfigureAwait(false))
```

Comment:
In my comment before i shared a concern about my lack of understanding regarding what happen with the abandoned tasks. I still don't know what happen but now we have this code that cancel them just in case they are not collected by the GC automatically after going out of scope.

I really don't like this. I would prefer the previous version because when in doubt the best is doing nothing.

Reply:
I did it because of this: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Threading/SemaphoreSlim.cs#L718 to allow the task clean itself because one can see that Tasks require completed state to dispose themselves: https://source.dot.net/#System.Private.CoreLib/Task.cs,1586

So the solution makes sense to me.

Reply:
@lontivero So I have consulted this solution with some .net guys and as far as I have found out, it should be correct. 

You can make it shorter like this:

```csharp
var linkedTcs = _tcs; // Copy reference so it cannot change.

if (linkedTcs.Task == await Task.WhenAny(linkedTcs.Task, Task.Delay(Period, stoppingToken)).ConfigureAwait(false))
{
  cts.Cancel(); // Ensure that the Task.Delay task is cleaned up.
}
```

but then garbage collection for `Task.Delay(Period, stoppingToken)` takes `Period` or more time. So it wastes resources a bit, but given the use of `PeriodicRunner` it is acceptable. Bigger strain on resources would be to use `PeriodicRunner` with `Period = 30 minutes` and then always triggering next round, then you would consume more and more memory without giving GC a chance to dispose those tasks in timely fashion.

I prefer the current version as it does not make any additional assumptions on how a person can use it.